packages/ui-react/src/containers/ModalProvider/useModal.ts

```diff
@@ -0,0 +1,49 @@
+import { type RefObject, useContext, useEffect, useRef } from 'react';
+import useEventCallback from '@jwp/ott-hooks-react/src/useEventCallback';
+
+import { ModalContext } from './ModalProvider';
+
+type Params = {
+  open: boolean;
+  onClose?: () => void;
+  modalRef: RefObject<HTMLElement>;
+  containerRef?: RefObject<HTMLElement>;
+};
+
+export const useModal = ({ open, onClose, modalRef, containerRef }: Params) => {
+  const modalId = useRef(String(Math.round(Math.random() * 1_000_000))).current;
+  const { openModal, closeModal, modals } = useContext(ModalContext);
+  const isOpen = modals.some((modal) => modal.modalId === modalId);
+
+  // replace with React `useEffectEvent` when stable: https://react.dev/reference/react/experimental_useEffectEvent
+  const openModalEvent = useEventCallback(() => {
+    if (open) {
+      openModal(modalId, modalRef, onClose, containerRef);
+    } else if (isOpen) {
+      closeModal(modalId, false);
+    }
+  });
+  const closeModalEvent = useEventCallback(() => {
+    if (isOpen) {
+      closeModal(modalId, false);
+    }
+  });
+
+  useEffect(() => {
+    // react when the `open` prop changes
+    openModalEvent();
+  }, [open, openModalEvent]);
```

Comment:
While creating the ModalManager, at one point, I had it exactly like that. But after a while, I felt that most logic in the ModalProvider became much more complex due to the following:

- The order of opening matters. To close the last opened modal, the stack needs to be sorted or time-stamped
- Currently, it's a bit funky that both states control the `open` state, I'd prefer to have a single state responsible for the open state, but this would mean a big refactor of the internal state and click actions (it's doable though)
- When we add the possibility to open modals using the context, we need to pass an `onOpen` callback so the state can be updated

Since we currently don't really need the benefit of opening modals using the context, I would like to keep this PR simple.

Reply:
I was thinking about having two stacks: 1 for registered windows with ids (or simply dictionary), 1 for opened windows. This way second one will behave the same way it does now. I am not sure about `onOpen` callback. In case we use `openModal`, then `isOpen` will be updated automatically.

My main concern is actually this part where we use several open states:

```
  const isOpen = modals.some((modal) => modal.modalId === modalId);

  // replace with React `useEffectEvent` when stable: https://react.dev/reference/react/experimental_useEffectEvent
  const openModalEvent = useEventCallback(() => {
    if (open) {
      openModal(modalId, modalRef, onClose, containerRef);
    } else if (isOpen) {
      closeModal(modalId, false);
    }
  });
```

and this one where we allow control via incoming props:

```
  useEffect(() => {
    // react when the `open` prop changes
    openModalEvent();
  }, [open, openModalEvent]);
```

So we have different ways now of how we close and open modals:

1. We open them using `open` prop.
2. We close them using callbacks (both for open and close). And we can't do it by settings `open` to false.

Should we just use open / show prop to control the state? This way no methods will be exposed from the hook.